news/_posts/performance/2021-01-27-PHP-scalability-and-filesystem.md

```diff
@@ -0,0 +1,231 @@
+
+But PHP allowing us to cache those calls and their responses since version 5.3, we'll make good use of this option and enable it as such:
+
+```
+realpath_cache_size = 4096K
+realpath_cache_ttl = 600
+```
+
+As usual, you may need to adjust those values depending on your own system and application (path length, number of files, etc..).
+
+### The open_basedir dilemma
+
+Don't forget that the real path cache directives are incompatible with the `open_basedir` one. Hence, you won't be able to enable both of them. Which could be a conundrum as one directive enables performance when the other enforces security.
+
+Our best advice would be to work on both sides by:
+
+* Either with containers (at PHP or kernel level), or VMs or whichever technology you're the most confortable with, working on security by minimizing the directories and contents the application can access to
+* Enabling the realpath cache directive in order to improve the application's or system's performances
+
+As always, we're aware our advice can not take several parameters into account - such as the security requirements of your application or system, your access to different platform architectures types or the resources at your disposal.
+
+Security is always a trade-off (performance, accessibility, ease of use: you name it).
+
+## opcache
+
+Opcache is another cache system available to us through PHP.
+
+This time we are not caching the filesystem paths, files and directories but the files contents themselves. And more precisely: the compiled files' contents, stored *in memory*. Which will minimize as much as possible the filesystem's accesses.
+
+This is a real booster for your system and/or application, here are some suggested configuration parameters:
+
+```
+[opcache]
+opcache.enable=1
+opcache.enable_cli=0
+opcache.memory_consumption=256
+opcache.interned_strings_buffer=32
+opcache.max_accelerated_files=16229
+opcache.max_wasted_percentage=10
+opcache.revalidate_freq=10
+opcache.fast_shutdown=1
+opcache.enable_file_override=0
+opcache.max_file_size=0
+```
+
+If you can manage it, here are few options you could also configure:
+
+```
+opcache.validate_timestamps=0
+opcache.revalidate_path=0
+```
+
+Keep in mind that, if you deactivate `validate_timestamps`, OPCache will never update your code, unless you trigger it (either through internal functions or by restarting the web server). Which will impact your deployment method if you're dealing with VMs and physical servers - but be irrelevant when working with containers.
+
+## Results
+
+So, what can you expect from a bit of caching, when trying to minimize the filesystem impact on PHP?
+
+Well, we're glad you ask!
+
+![TuningUsersCount][TuningUsersCount]
```

Comment:
I think we can improve the [alternative text](https://dev.to/stephencweiss/markdown-image-titles-and-alt-text-5fi1) 😉  for people with low network.

Reply:
Will do ;)